### PR TITLE
[nn] cross_entropy: no batch dim support

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference: #60585

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345

### CODE REVIEW

File: torch/nn/modules/loss.py [hunk 1]
```diff
@@ -1119,14 +1119,14 @@ class probabilities only when a single class label per minibatch item is too res
             `Rethinking the Inception Architecture for Computer Vision <https://arxiv.org/abs/1512.00567>`__. Default: :math:`0.0`.
 
     Shape:
-        - Input: :math:`(N, C)` where `C = number of classes`, or
+        - Input: :math:`(N, C)` or :math:`(C)` where `C = number of classes`, or
```
Comment (jbschlosser) [inline torch/nn/modules/loss.py:1122]: The docs here are getting more complex with the added cases, so I think it's important we strive for clarity. Wdyt about following the doc template for [`RNN`](https://pytorch.org/docs/1.9.1/generated/torch.nn.RNN.html)? That is, add a separate section starting with `where:` and list the shapes referenced. This could look something like this (formatting left out, sorry):

```
Input: Shape (C), (N, C), or (N, C, d_1, d_2, ..., d_K) with K >= 1 for K-dimensional loss.
Target: If containing class indices, shape (), (N), or (N, d_1, d_2, ..., d_K) with K >= 1 for K-dimensional loss; each value should be in [0, C-1]. If containing class probabilities, same shape as the input; each value should be in [0, 1].
Output: If reduction is 'none', same shape as the target. Otherwise, scalar.

where:
    C = number of classes
    N = batch size
```

Similar changes for the functional form in `torch/nn/functional.py`.

**Review by jbschlosser — COMMENTED** [review]
Implementation and tests look good! Added one comment below regarding docs. I think it's important we get the wording right since the supported shapes are getting complex

File: torch/nn/modules/loss.py (hunk 1 above)
Comment (kshitij12345) [reply torch/nn/modules/loss.py:1122]: Sure that makes sense. Thanks!

File: torch/nn/functional.py [hunk 2]
```diff
@@ -2895,14 +2895,8 @@ def cross_entropy(
...
-        input (Tensor) : :math:`(N, C)` where `C = number of classes` or :math:`(N, C, H, W)`
-            in case of 2D Loss, or :math:`(N, C, d_1, d_2, ..., d_K)` where :math:`K \geq 1`
-            in the case of K-dimensional loss. `input` is expected to contain unnormalized scores
-            (often referred to as logits).
-        target (Tensor) : If containing class indices, shape :math:`(N)` where each value is
-            :math:`0 \leq \text{targets}[i] \leq C-1`, or :math:`(N, d_1, d_2, ..., d_K)` with
-            :math:`K \geq 1` in the case of K-dimensional loss. If containing class probabilities,
-            same shape as the input.
+        input (Tensor) : Input Tensor (look below for valid shape).
+        target (Tensor) : Target Tensor (look below for valid shapes).
```
Comment (jbschlosser) [inline torch/nn/functional.py:2899]: nit: I think we should keep the high-level descriptions of input / target here even if we move shape info below.
Ex:
```
input (Tensor): Predicted unnormalized scores (often referred to as logits); see Shape section below for supported shapes.
target (Tensor): Ground truth class indices or class probabilities; see Shape section below for supported shapes.
```

File: torch/nn/functional.py [hunk 3]
```diff
@@ -2930,6 +2924,22 @@ def cross_entropy(
             become a mixture of the original ground truth and a uniform distribution as described in
             `Rethinking the Inception Architecture for Computer Vision <https://arxiv.org/abs/1512.00567>`__. Default: :math:`0.0`.
 
+    Shape:
+        - Input: Shape :math:`(C)`, :math:`(N, C)` or :math:`(N, C, d_1, d_2, ..., d_K)` with :math:`K \geq 1`
+          in the case of `K`-dimensional loss.
+        - Target: If containing class indices, shape :math:`()` or :math:`(N)` where each value is
+          between [0, C), otherwise :math:`(N, d_1, d_2, ..., d_K)` with
+          :math:`K \geq 1` in the case of K-dimensional loss with same shape as the input and
+          each value should be between [0, 1].
```
Comment (jbschlosser) [inline torch/nn/functional.py:2933]: Similar comment here as below regarding correctness of target docs

**Review by jbschlosser — COMMENTED** [review]
Pretty close to wrapping this up; few nitpicky comments for docs :)

File: torch/nn/modules/loss.py [hunk 4]
```diff
@@ -1119,16 +1119,21 @@ class probabilities only when a single class label per minibatch item is too res
...
-          :math:`0 \leq \text{targets}[i] \leq C-1`, or :math:`(N, d_1, d_2, ..., d_K)` with
-          :math:`K \geq 1` in the case of K-dimensional loss. If containing class probabilities,
-          same shape as the input.
-        - Output: If :attr:`reduction` is ``'none'``, shape :math:`(N)` or
-          :math:`(N, d_1, d_2, ..., d_K)` with :math:`K \geq 1` in the case of K-dimensional loss.
-          Otherwise, scalar.
+        - Target: If containing class indices, shape :math:`()` or :math:`(N)` where each value is
+          between [0, C), otherwise :math:`(N, d_1, d_2, ..., d_K)` with
+          :math:`K \geq 1` in the case of K-dimensional loss with same shape as the input and
+          each value should be between [0, 1].
```
[the remark this answers is not in the document]
Comment (kshitij12345) [reply torch/nn/modules/loss.py]: Ah. I was mistaken about the shapes. Thanks for clarifying. Have updated the docs.

**Comment by kshitij12345** [comment]
Updated Docs (from CI):
https://docs-preview.pytorch.org/71055/generated/torch.nn.CrossEntropyLoss.html?highlight=cross%20entropy#torch.nn.CrossEntropyLoss

https://docs-preview.pytorch.org/71055/generated/torch.nn.functional.cross_entropy.html?highlight=cross%20entropy#torch.nn.functional.cross_entropy

**Review by jbschlosser — COMMENTED** [review]
Just did a read-through of the rendered docs and realized input and target shapes are specified twice in the doc, once in the long mathy part. Once that part is updated, we should be good to go!

File: torch/nn/modules/loss.py [hunk 5]
```diff
@@ -1119,16 +1119,20 @@ class probabilities only when a single class label per minibatch item is too res
             `Rethinking the Inception Architecture for Computer Vision <https://arxiv.org/abs/1512.00567>`__. Default: :math:`0.0`.
 
     Shape:
-        - Input: :math:`(N, C)` where `C = number of classes`, or
-          :math:`(N, C, d_1, d_2, ..., d_K)` with :math:`K \geq 1`
+        - Input: Shape :math:`(C)`, :math:`(N, C)` or :math:`(N, C, d_1, d_2, ..., d_K)` with :math:`K \geq 1`
```
[the remark this answers is not in the document]
Comment (kshitij12345) [reply torch/nn/modules/loss.py:1122]: I should have noticed them and the above rendering issue. Thanks for being patient :)!

**Review by jbschlosser — APPROVED** [review]
LGTM! Thanks for the update :)
